New scalable routing datapath

tgraf opened the pull request:
Scale testing beyond 1k nodes have lead to the conclusion that the interactions
with the Linux routing tables have been inefficient. Several small factors are
contributing to the overhead with the main reason behind naive routing updates
for the entire list of nodes on certain node events.

## Summary of changes

 * Separation of node definition, node management and datapath implementation
   based on node events.

 * pkg/node contains the definition of a node, its addressing, logic to
   determine the primary external address, determination of local hostname, and
   management of the addressing for the local node. The addressing should be
   moved out of this package in a future PR but was outside of thes scope of
   this commit.

 * pkg/node/manager manages the list of nodes and implements the logic to
   handle node events from multiple sources. Sources include:

    * local node definition
    * local agent information
    * kubernetes
    * kvstore

   Two new metrics are introduced:

   * gauge cilium_nodes_all_num
   * counter cilium_nodes_all_events_received_total {"all", "update", "delete"}

   This allows to visualize the current cluster size as well as the rate of
   node updates received per minute.

   The manager periodically requests the datapath to validate the
   implementation of a node. As the cost of validation grows with the cluster
   size, the interval in which the validation is performed is made dependent on
   the cluster size using a logarithmic function.

 * pkg/datapath is extended to contain the requierments for a datapath specific
   implementation. This includes handling of node events, changes to the local
   node configuration and platform specific node addressing.

   These new types of interfaces should be used to abstract all datapath
   specific requirements. This commit introduces the interfaces for node event
   handling but the same concept can be applied to all datapath interactions
   and thus allow to provide comprehensive isolated unit tests for the datapath
   interactions.

 * pkg/datapath/fake is a fake implementation of the datapath for testing

 * pkg/datapath/linux implements the Linux specific datapath portion for node
   events handling according to the node configuration. This includes
   determination on what type of routes to push down based on enablement of
   address families, encapsulation, direct-routing and use of single cluster
   prefixes.

   The datapath layer is responsible to optimize detection of changes in node
   information and only perform routing layer changes when the definition has
   changed. Changes include:

    * A node changing its IP address
    * A node starting or stopping to announce an allocation CIDR
    * A node changing its allocation CIDR

 * The existing AutoIPv6NodeRoutes mode is made more generic and automated. It
   now applies to IPv6 and IPv4 and no longer requires to specify any device
   information. When EnableAutoDirectRouting is enabled, the datapath layer
   will detect whether a direct routing connection is possible between two
   nodes and install a direct node route automatically. If the direct routing
   is not feasible, likely because the remote node does not share the same L2
   segment, fallback to encapsulation is performed.

## Benchmarking

Cost of node update event when node information is not changing:

```
BenchmarkNoChangeNodeUpdate                       5000000            628 ns/op
BenchmarkNoChangeNodeUpdateDirectRouteAll         5000000            699 ns/op
BenchmarkNoChangeNodeUpdateEncapAll                500000           3637 ns/op
```

Cost of node update event when node CIDR is changing. This is the cost paid
when a node is added, the node IP changes or the node changes its CIDR.

```
BenchmarkNodeUpdateAll                            5000000            639 ns/op
BenchmarkNodeUpdateIPv4                           5000000            618 ns/op
BenchmarkNodeUpdateIPv6                           5000000            662 ns/op
BenchmarkNodeUpdateDirectRouteAll                    1000        1131640 ns/op
BenchmarkNodeUpdateDirectRouteIPv4                   5000         557249 ns/op
BenchmarkNodeUpdateDirectRouteIPv6                   5000         661931 ns/op
BenchmarkNodeUpdateEncapAll                          1000        1915020 ns/op
BenchmarkNodeUpdateEncapIPv4                         1000        1007195 ns/op
BenchmarkNodeUpdateEncapIPv6                         1000        1374214 ns/op
BenchmarkNodeUpdateEncapSingleClusterRouteAll      200000          12388 ns/op
BenchmarkNodeUpdateEncapSingleClusterRouteIPv4     200000           7680 ns/op
BenchmarkNodeUpdateEncapSingleClusterRouteIPv6     200000           6625 ns/op
```

Cost of validation to determine whether a node is correctly implemented in the
datapath. This is the cost paid per node via the controller to periodically
check the datapath implementation.

```
BenchmarkNodeValidateImplementationAll                        2000         813551 ns/op
BenchmarkNodeValidateImplementationIPv4                       2000         839429 ns/op
BenchmarkNodeValidateImplementationIPv6                       2000         980503 ns/op
BenchmarkNodeValidateImplementationDirectRouteAll             2000         861470 ns/op
BenchmarkNodeValidateImplementationDirectRouteIPv4            2000         552912 ns/op
BenchmarkNodeValidateImplementationDirectRouteIPv6            5000         487668 ns/op
BenchmarkNodeValidateImplementationEncapAll                   1000        1286646 ns/op
BenchmarkNodeValidateImplementationEncapIPv4                  5000         722960 ns/op
BenchmarkNodeValidateImplementationEncapIPv6                  2000         791391 ns/op
BenchmarkNodeValidateImplementationEncapSingleClusterAll      5000         365746 ns/op
BenchmarkNodeValidateImplementationEncapSingleClusterIPv4     5000         377625 ns/op
BenchmarkNodeValidateImplementationEncapSingleClusterIPv6     5000         427404 ns/op
```

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/6617)
<!-- Reviewable:end -->

eloycoto commented on test/k8sT/DatapathConfiguration.go:
```diff
@@ -65,99 +71,69 @@ var _ = Describe("K8sTunnelTest", func() {
+		validateBPFTunnelMap := func() {
 			By("Checking that BPF tunnels are in place")
+			ciliumPod, err := kubectl.GetCiliumPodOnNode(helpers.KubeSystemNamespace, helpers.K8s1)
+			ExpectWithOffset(1, err).Should(BeNil())
```
assert message here please.

eloycoto commented on test/k8sT/DatapathConfiguration.go:
```diff
@@ -65,99 +71,69 @@ var _ = Describe("K8sTunnelTest", func() {
+		It("Check connectivity with VXLAN encapsulation", func() {
+			deployCilium("cilium-ds-patch-vxlan.yaml")
+			validateBPFTunnelMap()
+			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue())
```
If you can add a assert message will be great.

eloycoto commented on test/k8sT/DatapathConfiguration.go:
```diff
@@ -65,99 +71,69 @@ var _ = Describe("K8sTunnelTest", func() {
+	Context("DirectRouting", func() {
+		It("Check connectivity with automatic direct nodes routes", func() {
+			deployCilium("cilium-ds-patch-auto-node-routes.yaml")
+			Expect(testPodConnectivityAcrossNodes(kubectl)).Should(BeTrue())
```
assert message please.

aanm commented on daemon/daemon_main.go:
```diff
@@ -325,8 +326,9 @@ func init() {
-	flags.Bool(option.AutoIPv6NodeRoutesName, false, "Automatically adds IPv6 L3 routes to reach other nodes for non-overlay mode (--device) (BETA)")
-	option.BindEnv(option.AutoIPv6NodeRoutesName)
+	flags.Bool(option.LegacyAutoIPv6NodeRoutesName, false, "Automatically adds IPv6 L3 routes to reach other nodes for non-overlay mode (--device) (BETA)")
+	flags.MarkHidden(option.LegacyAutoIPv6NodeRoutesName)
```
use `flags.MarkDeprecated` as well

aanm commented on daemon/daemon_main.go:
```diff
@@ -973,8 +978,12 @@ func initEnv(cmd *cobra.Command) {
+
 	log.Info("Initializing daemon")
-	d, restoredEndpoints, err := NewDaemon()
+	d, restoredEndpoints, err := NewDaemon(linuxdatapath.NewDatapath(datapathConfig))
```
What's with this `NewDatapath`? Why can't it be contained in its own package? The `HostDevice` could be derived inside that package.

tgraf replied:
The purpose of the Datapath interface is to allow testing with a fake datapath (`pkg/datapath/fake`) which allows to do unit testing of all the infrastructure on top of the datapath without requiring privileged tests. The node manager tests is what is using this right now. Next logical tests will be the entire DaemonTestSuite and endpoint regeneration tests so we can test it without requiring performing datapath operations.

The HostDevice is configured outside so the datapath module can do unit testing with randomly generated host device names and we can even simulate two datapaths on a single machine in a unit test and test route interactions as done for the direct routing tests.

aanm commented on daemon/health.go:
```diff
@@ -37,19 +37,11 @@ func (d *Daemon) initHealth() {
-	if err != nil {
-		log.WithError(err).Fatal("Error while set health IPv6 ip on the local node.")
-	}
+	d.setIPv4HealthIP(health4)
```
Why the change?

tgraf replied:
The ownership of the local `node.Node` instance is moved from a global variable into the agent. The only global variables that remain in `pkg/node` are `pkg/node/node_address.go` which, as written in the commit message, can be converted next. The new datapath already uses a `NodeAddressing` interface so the change is becoming very simple. This again allows to use fake node addressing in the datapath unit testing.

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -0,0 +1,530 @@
+// with encapsulation mode enabled. The CIDR and IP of both the old and new
+// node are provided as context. The caller expects the tunnel mapping in the
+// datapath to be updated.
+func updateTunnelMapping(oldCIDR, newCIDR *net.IPNet, oldIP, newIP net.IP, firstAddition, enabled bool) {
```
Please change the argument `enabled` with `encapEnabled`

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -0,0 +1,530 @@
+	return netlink.RouteListFiltered(family, routeSpec, netlink.RT_FILTER_DST|netlink.RT_FILTER_GW|netlink.RT_FILTER_OIF)
+}
+
+func (n *linuxNodeHandler) updateDirectRoute(oldCIDR, newCIDR *net.IPNet, oldIP, newIP net.IP, firstAddition, enabled bool) error {
```
same here for the `enable` argument

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -0,0 +1,530 @@
+	return netlink.RouteListFiltered(family, routeSpec, netlink.RT_FILTER_DST|netlink.RT_FILTER_GW|netlink.RT_FILTER_OIF)
+}
+
+func (n *linuxNodeHandler) updateDirectRoute(oldCIDR, newCIDR *net.IPNet, oldIP, newIP net.IP, firstAddition, enabled bool) error {
```
Are there unit tests for this function?

tgraf replied:
Yes,

```
func (s *linuxPrivilegedTestSuite) TestNodeUpdateDirectRouting(c *check.C) {
```

aanm commented on pkg/datapath/linux/node.go:
```diff
@@ -0,0 +1,530 @@
+	delete(n.nodes, oldNode.Identity())
+
+	if n.isInitialized {
+		return n.nodeDelete(&oldNode)
```
Small optimization, this line should only be called if `n.nodes` contained the `oldNode` in the map

aanm commented on pkg/datapath/linux/route/route_linux.go:
```diff
@@ -92,6 +115,7 @@ func ipFamily(ip net.IP) int {
 func lookup(link netlink.Link, route *netlink.Route) *netlink.Route {
 	routes, err := netlink.RouteList(link, ipFamily(route.Dst.IP))
 	if err != nil {
+		fmt.Printf("NO LINK\n")
```
`NO PRINTF`

aanm commented on pkg/datapath/linux/route/route_linux.go:
```diff
@@ -190,7 +239,21 @@ func replaceRoute(route Route, mtuConfig mtu.Configuration) (bool, error) {
 
 	if lookup(link, &routeSpec) == nil {
-		if err := netlink.RouteReplace(&routeSpec); err != nil {
+		err := fmt.Errorf("routeReplace not called yet")
```
On which occasion would this exact message occur?

tgraf replied:
Never. I figured that it is better to have a unique message appear that we can grep for *if* it ever happens than nil which makes it impossible to debug.

aanm commented on pkg/node/manager/manager.go:
```diff
@@ -0,0 +1,347 @@
+
+	nodes := make(map[node.Identity]node.Node)
+	for nodeIdentity, entry := range m.nodes {
+		entry.mutex.Lock()
```
What is this mutex protecting? I looked through all code and doesn't seem to do anything relevant as its access is already being protected by `m.mutex`

tgraf replied:
The documentation section of this mutex says:
```
        // mutex serves two purposes:
        // 1. Serialize any direct access to the node field in this entry.
        // 2. Serialize all calls do the datapath layer for a particular node.
        //
        // See description of Manager.mutex for more details
        //
        // If both the nodeEntry.mutex and Manager.mutex must be held, then the
        // Manager.mutex must *always* be acquired first.
        mutex lock.Mutex
```

It fulfills point 1 for consistency. It is not strictly required as we don't unlock Manager.mutex in this case but it is good to be consistent here and make sure that the documented locking rules are always followed.

aanm commented on pkg/node/node.go:
```diff
@@ -309,20 +294,12 @@ func (n *Node) Marshal() ([]byte, error) {
 		return err
 	}
 
 	*n = newNode
```
This still needs to be solved to fix https://github.com/cilium/cilium/issues/6572

tgraf replied:
ACK. I originally had the fix in but opted to keep it outside so we can push a fix that we can also backport to 1.3

aanm commented on pkg/node/store/store.go:
```diff
@@ -39,17 +40,23 @@ var (
-		n.OnUpdate()
+		nodeCopy := *n
+		nodeCopy.Source = node.FromKVStore
+		o.manager.NodeUpdated(nodeCopy)
```
Be aware this is not a copy, the internal slices of `n` were not copied. Please use DeepCopy

aanm commented on pkg/node/store/store.go:
```diff
@@ -39,17 +40,23 @@ var (
-		n.OnDelete()
+		nodeCopy := *n
+		nodeCopy.Source = node.FromKVStore
+		o.manager.NodeDeleted(nodeCopy)
```
Same here for be aware this is not a copy, the internal slices of `n` were not copied. Please use DeepCopy

aanm commented on pkg/node/store/store.go:
```diff
@@ -59,14 +66,14 @@ type NodeRegistrar struct {
 
 // RegisterNode registers the local node in the cluster
-func (nr *NodeRegistrar) RegisterNode(n *node.Node) error {
+func (nr *NodeRegistrar) RegisterNode(n *node.Node, manager *nodemanager.Manager) error {
```
We are slowly moving again to all packages depending on all other packages. Consider writing an interface for `nodemanager.Manager` instead so that `NodeRegistrar` does not import `nodemanager` package.

tgraf replied:
Very good point. Done

aanm commented on pkg/option/config.go:
```diff
@@ -899,6 +905,14 @@ func (c *DaemonConfig) Populate() {
+	// automatic direct routing. Unlike the old option, it will also enable
+	// direct routing for IPv4. Better than breaking the option. The old
+	// option was not frequently used so this addition in scope is fine.
+	if viper.GetBool(LegacyAutoIPv6NodeRoutesName) {
```
Since this is a legacy why does it can overwrite the new option? Shouldn't `c.EnableAutoDirectRouting` default to  `viper.GetBool(LegacyAutoIPv6NodeRoutesName)` and if `viper.GetBool(EnableAutoDirectRoutingName)` is enabled that the option should be overwritten?

tgraf replied:
We want people to *remove* the legacy option in order to turn off the feature, we don't want them to add EnableAutoDirectRoutingName=false in addition. Does that make sense?

joestringer commented on pkg/node/manager/manager.go:
```diff
@@ -0,0 +1,347 @@
+// node in the manager is added or updated if the source is allowed to update
+// the node. If an update or addition has occured, NodeUpdate() of the datapath
+// interface is invoked.
+func (m *Manager) NodeUpdated(n node.Node) {
```
Minor nits: Seems like we call this operation 'Upsert' in other parts of the codebase. Also, using past tense in function names is a little confusing because the callers are named things like `updateK8sNodeTunneling` and the callees `NodeUpdate`.

(Also, this is the nodemanager. `d.nodeManager.NodeUpdated()` stutters.)

tgraf replied:
It's meant as an event handler. The caller won't know whether the node will be replaced in the manager or not. That's why I don't like `Upsert()` as it implies the caller is in charge.

joestringer commented on pkg/datapath/linux/route/route_linux.go:
```diff
@@ -106,7 +129,7 @@ func lookup(link netlink.Link, route *netlink.Route) *netlink.Route {
 		aMaskLen, aMaskBits := r.Dst.Mask.Size()
 		bMaskLen, bMaskBits := route.Dst.Mask.Size()
-		if r.LinkIndex == route.LinkIndex && r.Scope == route.Scope &&
+		if r.Scope == route.Scope &&
```
Was the LinkIndex check wrong here, or unnecessary?

tgraf replied:
Yes, this prevented the lookup from working. I will isolate the bugfix as we might want to backport it.

tgraf replied:
This fix has been factored out with commit description

joestringer commented on pkg/datapath/node.go:
```diff
@@ -0,0 +1,118 @@
+	// points the allocation prefix of the local node. Use of this option
+	// is useful when another component is responsible for the routing of
+	// the allocation CIDR IPs into Cilium endpoints.
+	DisableLocalNodeRoute bool
```
nit: Is there a reason to set this false by default and check `!DisableLocalNodeRoute`? This would be easier to read if it were `EnableLocalNodeRoute` like the other options above.

tgraf replied:
Not really, changed

joestringer commented on pkg/node/node.go:
```diff
@@ -20,9 +20,33 @@ import (
+	FromAgentLocal Source = "agent-local"
+
+	// FromLocalNode is the source used for the local node
+	FromLocalNode Source = "local-node"
```
What's the difference between `FromAgentLocal` and `FromLocalNode`?

Is `FromAgentLocal` here the same as the `FromAgentLocal` under `pkg/ipcache`?

tgraf replied:
It's two different sources, before it was possible for the kvstore to overwrite the local node entry in the manager if node names overlapped.

joestringer commented on pkg/datapath/linux/node.go:
```diff
@@ -0,0 +1,533 @@
+		// When the protocol family is disabled, the initial node addition will
+		// trigger a deletion to clean up leftover entries. The deletion happens
+		// in quiet mode as we don't know whether it exists or not
+		if newCIDR != nil && firstAddition {
```
Why do we care if it's the firstAddition? Is it to avoid dumping routes later on? But then I would expect the route list to be empty after the first time, in which case it's perhaps 1 syscall saved in the subsequent cases?

Also, `deleteDirectRoute()` will check if `newCIDR == nil` and become no-op in that case.

(If this feedback leads to any changes, then it presumably also applies to the tunnel route case too.)

tgraf replied:
This handles changing in configuration option. If the agent is running with direct routing and is then restarted with direct routing disabled, the first node addition of any node added will remove the old route. Blind deletions are quite costly so it's done once when the node is added for the first time.

joestringer replied:
If I followed correctly, the callee code does a dump of routes first then attempts deletion for routes that match, rather than a blind deletion. But your comment may still apply; I don't have a sense for how expensive that is in practice.

tgraf replied:
It is a lot more costly than what I would have guessed because it involves:
- link lookup
- route list

The route list in the kernel ist mostly sequential it seems. I measured the lookup with a couple of routes first and it was OKish but when I added 10k routes to simulate 5k nodes it became a lot more expensive, hence the optimization here.

There is obviously more we can do such as caching link lookups or listening to Netlink notifications. What is PR does is good enough for 5K nodes with some optimization potential left.

joestringer commented on pkg/datapath/linux/node.go:
```diff
@@ -0,0 +1,533 @@
+	// Node allocation CIDR has changed
+	case oldCIDR != nil && newCIDR != nil && !oldCIDR.IP.Equal(newCIDR.IP):
+		deleteTunnelMapping(oldCIDR, false)
+	}
```
The switch statement conditions are very similar to open-coding `cidrNodeMappingUpdateRequired()`. Can it be used?

tgraf replied:
The tunnel entry may not be deleted if only the node IP is different.

joestringer commented on pkg/datapath/linux/node.go:
```diff
@@ -0,0 +1,533 @@
+	// Node allocation CIDR has changed
+	case oldCIDR != nil && newCIDR != nil && !oldCIDR.IP.Equal(newCIDR.IP):
+		n.deleteDirectRoute(oldCIDR, oldIP)
+	}
```
This switch statement's conditions appear to directly open-code `cidrNodeMappingUpdateRequired()`. Can it be re-used?

tgraf replied:
The main difference in the logic is newly announced CIDR vs no longer announcing a CIDR. I don't really mind what we do, I'm not sure one is much clearer over the other that's why I've added comments to explain each condition.

joestringer reviewed (approved):
I'm excited about mocking out the datapath, and the idea of moving from refreshing all routes any time a node is added or removed to only handling the routes for the node that's changing looks good. The improvements to the option handling for the different modes are also a positive step.

I found it quite difficult to review this as refactoring was included in the same patch as modifying behaviour, and that patch is quite large. On the plus side, there's a bunch of new unit tests for the functionality so that suggests that the new configuration should be in a reasonably good state.

Most of my feedback below is random nits around naming, code reuse, minor simplifications. I see some potential for further code reuse between the local node routes for tunnels and direct routes in terms of deletion, update functions, etc but it's hard to tell if there's any meaningful discrepancy between the codepaths there.

ianvernon commented on daemon/daemon.go:
```diff
@@ -168,8 +171,14 @@ type Daemon struct {
-	policyTrigger *trigger.Trigger
+	mtuConfig      mtu.Configuration
+	policyTrigger  *trigger.Trigger
+	nodeManager    *nodemanager.Manager
```
This is stylistic, but it'd be nice to group the `nodeManager`, `nodeConfig`, `nodeRegistrar`, `nodeRegistered`, `localNode`, into a type within the daemon to reduce the amount of fields within the daemon / more logically collocate node-related fields for daemon.

tgraf replied:
I've introduced `nodeDiscovery` type which does this

ianvernon commented on daemon/k8s_watcher.go:
```diff
@@ -1816,19 +1806,20 @@ func (d *Daemon) updateK8sNodeV1(k8sNodeOld, k8sNodeNew *v1.Node) error {
 func (d *Daemon) deleteK8sNodeV1(k8sNode *v1.Node) error {
+	oldNode := k8s.ParseNode(k8sNode, node.FromKubernetes)
+	// Ignore own node
+	if oldNode.Name == node.GetName() {
```
Is this ever possible?

tgraf replied:
Of course

ianvernon commented on daemon/local_node.go:
```diff
@@ -0,0 +1,88 @@
+		select {
+		case <-d.nodeRegistered:
+		case <-time.NewTimer(defaults.NodeInitTimeout).C:
+			log.Fatalf("Unable to initialize local node due timeout")
```
Nit: due to (I think this was there before, so you don't have to fix it as part of this PR)

ianvernon commented on pkg/datapath/datapath.go:
```diff
@@ -0,0 +1,27 @@
+// Datapath is the interface to abstract all datapath interactions. The
+// abstraction allows to implement the datapath requirements with multiple
+// implementations
+type Datapath interface {
```
hooray!!!

ianvernon commented on pkg/datapath/linux/datapath.go:
```diff
@@ -0,0 +1,55 @@
+// configuration cannot change throughout the lifetime of a datapath object.
+type DatapathConfiguration struct {
+	// HostDevice is the name of the device to be used to access the host.
+	HostDevice string
```
If you want to ensure that this will never be changed throughout the lifetime of the datapath object, you could make it private and have setters that return an error if it is tried to be changed after being set (i.e., the value of the field is its zero value, in this case "").

tgraf replied:
DatapathConfiguration  is passed in as a copy into NewDatapath so it is already impossible to ever change this

ianvernon commented on pkg/datapath/linux/datapath_test.go:
```diff
@@ -1,4 +1,4 @@
-// Copyright 2018 Authors of Cilium
+// Copyright 2019 Authors of Cilium
```
Change to 2018-2019

tgraf replied:
This is new code, diff just abuses a random old file

ianvernon reviewed (approved):
Approving for docs. 

I didn't look at all of the golang code, because this was really hard to review due to the sheer size of the PR. It'd be really helpful to split out interface definitions into separate commits, deletion of code into separate commits, etc. to make this more digestible in the future. This is hard to do post-coding, but I think it is worth the cycles to do this.

tgraf committed:
test: Add test for automatic direct node routing

This restructures the tunnel tests into a more generic datapath configuration
test suite and adds a dedicated test for automatic direct node routes.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
cidr: Add cidr package to represent net.IPNet

Provide a cidr package to wrap net.IPNet so we can provide DeepCopy functions
for use of cidr.CIDR in structures which require deepcopy code generation.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
bpf: Use unique map names when running unit tests

The privileged unit tests shared the map names when performing compile and
loader testing. This caused live maps of a running Cilium instance to be
manipulated or cleared. Use defines to specify unit testing map names and
remove those maps when tests are done.

This allows running privileged unit tests side by side running Cilium
instances.

Fixes: #6698

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
datapath/linux: New scalable routing layer via Linux datapath implementation

Scale testing beyond 1k nodes have lead to the conclusion that the interactions
with the Linux routing tables have been inefficient. Several small factors are
contributing to the overhead with the main reason behind naive routing updates
for the entire list of nodes on certain node events.

pkg/datapath/linux implements the Linux specific datapath portion for node
events handling according to the node configuration. This includes
determination on what type of routes to push down based on enablement of
address families, encapsulation, direct-routing and use of single cluster
prefixes.

The datapath layer is responsible to optimize detection of changes in node
information and only perform routing layer changes when the definition has
changed. Changes include:

* A node changing its IP address
* A node starting or stopping to announce an allocation CIDR
* A node changing its allocation CIDR

The existing AutoIPv6NodeRoutes mode is made more generic and automated. It now
applies to IPv6 and IPv4 and no longer requires to specify any device
information. When EnableAutoDirectRouting is enabled, the datapath layer will
detect whether a direct routing connection is possible between two nodes and
install a direct node route automatically. If the direct routing is not
feasible, likely because the remote node does not share the same L2 segment,
fallback to encapsulation is performed.

Cost of node update event when node information is not changing:

```
BenchmarkNoChangeNodeUpdate                       5000000            628 ns/op
BenchmarkNoChangeNodeUpdateDirectRouteAll         5000000            699 ns/op
BenchmarkNoChangeNodeUpdateEncapAll                500000           3637 ns/op
```

Cost of node update event when node CIDR is changing. This is the cost paid
when a node is added, the node IP changes or the node changes its CIDR.

```
BenchmarkNodeUpdateAll                            5000000            639 ns/op
BenchmarkNodeUpdateIPv4                           5000000            618 ns/op
BenchmarkNodeUpdateIPv6                           5000000            662 ns/op
BenchmarkNodeUpdateDirectRouteAll                    1000        1131640 ns/op
BenchmarkNodeUpdateDirectRouteIPv4                   5000         557249 ns/op
BenchmarkNodeUpdateDirectRouteIPv6                   5000         661931 ns/op
BenchmarkNodeUpdateEncapAll                          1000        1915020 ns/op
BenchmarkNodeUpdateEncapIPv4                         1000        1007195 ns/op
BenchmarkNodeUpdateEncapIPv6                         1000        1374214 ns/op
BenchmarkNodeUpdateEncapSingleClusterRouteAll      200000          12388 ns/op
BenchmarkNodeUpdateEncapSingleClusterRouteIPv4     200000           7680 ns/op
BenchmarkNodeUpdateEncapSingleClusterRouteIPv6     200000           6625 ns/op
```

Cost of validation to determine whether a node is correctly implemented in the
datapath. This is the cost paid per node via the controller to periodically
check the datapath implementation.

```
BenchmarkNodeValidateImplementationAll                        2000         813551 ns/op
BenchmarkNodeValidateImplementationIPv4                       2000         839429 ns/op
BenchmarkNodeValidateImplementationIPv6                       2000         980503 ns/op
BenchmarkNodeValidateImplementationDirectRouteAll             2000         861470 ns/op
BenchmarkNodeValidateImplementationDirectRouteIPv4            2000         552912 ns/op
BenchmarkNodeValidateImplementationDirectRouteIPv6            5000         487668 ns/op
BenchmarkNodeValidateImplementationEncapAll                   1000        1286646 ns/op
BenchmarkNodeValidateImplementationEncapIPv4                  5000         722960 ns/op
BenchmarkNodeValidateImplementationEncapIPv6                  2000         791391 ns/op
BenchmarkNodeValidateImplementationEncapSingleClusterAll      5000         365746 ns/op
BenchmarkNodeValidateImplementationEncapSingleClusterIPv4     5000         377625 ns/op
BenchmarkNodeValidateImplementationEncapSingleClusterIPv6     5000         427404 ns/op
```

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
test; Report error when docker network cannot be created

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
route: move pkg/datapath/route to pkg/datapath/linux/route

This prepares the route utility layer for the upcoming pkg/datapath/linux
package. It's primarily a code move and renames the main update and delete
functio to be easier to undertand:

route.ReplaceRoute() -> route.Upsert()
Route.DeleteRoute() -> route.Delete()

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
datapath/fake: Add fake datapath implementation

fake implementation of the datapath for testing

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
test: Run privileged unit tests as part of runtime test

As the unit tests have become more advanced with the addition of a lot more
testing on routing layer interaction, running them in a container using
docker-compose is no longer possible. Use the runtime infrastructure instead.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
node: Use new routing datapath for node events

This commit achieves separation of node definition, node management and
datapath implementation based on node events.

pkg/node contains the definition of a node, its addressing, logic to determine
the primary external address, determination of local hostname, and management
of the addressing for the local node. The addressing should be moved out of
this package in a future PR but was outside of thes scope of this commit.

pkg/node/manager manages the list of nodes and implements the logic to handle
node events from multiple sources. Sources include:

    * local node definition
    * local agent information
    * kubernetes
    * kvstore

Two new metrics are introduced:

   * gauge cilium_nodes_all_num
   * counter cilium_nodes_all_events_received_total {"all", "update", "delete"}
   * counter cilium_nodes_all_datapath_validations_total

This allows to visualize the current cluster size as well as the rate of node
updates received per minute.

The manager periodically requests the datapath to validate the implementation
of a node. As the cost of validation grows with the cluster size, the interval
in which the validation is performed is made dependent on the cluster size
using a logarithmic function.

The local node management is moved into daemon/node_discovery.go and simplified
by allocating health IPs before starting node discovery. This avoids requiring
to annodate the k8s node multiple times.

Signed-off-by: Thomas Graf <thomas@cilium.io>

tgraf committed:
datapath: Introduce datapath interface abstraction

pkg/datapath is extended to contain the requierments for a datapath specific
implementation. This includes handling of node events, changes to the local
node configuration and platform specific node addressing.

These new types of interfaces should be used to abstract all datapath specific
requirements. This commit introduces the interfaces for node event handling but
the same concept can be applied to all datapath interactions and thus allow to
provide comprehensive isolated unit tests for the datapath interactions.

Signed-off-by: Thomas Graf <thomas@cilium.io>
